### PR TITLE
multiple signatures (rebased from vizicsaba89 PR)

### PR DESCRIPTION
**Pull request by andres-blanco**
It seems like @vizicsaba89 rebased the commits on develop: https://github.com/vizicsaba89/node-signpdf
I've made a fork here https://github.com/SIU-Toba/node-signpdf/ rebasing the changes from vizicsaba89 onto your updated develop branch.
I'm interested in this feature, so if there are any fixes to be done count me in.

### CODE REVIEW

**Comment by vbuch** [comment]
Without having looked into any of the code, I would ask for 100% coverage. The package had it so far, and I think 2.4% are feasible.

File: src/helpers/plainAddPlaceholder/getPagesDictionaryRef.js [hunk 1]
```diff
@@ -3,15 +3,20 @@ import SignPdfError from '../../SignPdfError';
 /**
  * @param {Object} info As extracted from readRef()
  */
-export default function getPagesDictionaryRef(info) {
-    const pagesRefRegex = new RegExp('\\/Type\\s*\\/Catalog\\s*\\/Pages\\s+(\\d+\\s\\d+\\sR)', 'g');
-    const match = pagesRefRegex.exec(info.root);
-    if (match === null) {
+const getPagesDictionaryRef = (info) => {
+    const catalogPosition = info.root.toString().indexOf('/Type/Catalog');
```
Comment (vbuch) [inline src/helpers/plainAddPlaceholder/getPagesDictionaryRef.js]: So, a dounf by @therpobinski, this peace brakes the function.

File: src/helpers/plainAddPlaceholder/index.js [hunk 2]
```diff
@@ -11,6 +11,13 @@ import createBufferRootWithAcroform from './createBufferRootWithAcroform';
 import createBufferPageWithAnnotation from './createBufferPageWithAnnotation';
 import createBufferTrailer from './createBufferTrailer';
 
+const isContainBufferRootWithAcrofrom = (pdf) => {
```
Comment (vbuch) [inline src/helpers/plainAddPlaceholder/index.js]: typo in the name of the function. Acrofrom should be Acrof**or**m

File: src/helpers/plainAddPlaceholder/index.js [hunk 3]
```diff
@@ -25,22 +32,27 @@ const plainAddPlaceholder = ({
     reason,
     signatureLength = DEFAULT_SIGNATURE_LENGTH,
 }) => {
+    if (!(pdfBuffer instanceof Buffer)) {
+        console.log(pdfBuffer);
```
Comment (vbuch) [inline src/helpers/plainAddPlaceholder/index.js]: No console.log in production code, please.

File: src/helpers/plainAddPlaceholder/readPdf.js [hunk 4]
```diff
@@ -16,37 +16,37 @@ const readPdf = (pdfBuffer) => {
...
+
+    if (refTable.maxOffset > xRefPosition) {
         throw new SignPdfError(
-            'The document already contains a form. This is not yet supported.',
+            'Ref table is not at the end of the document. This document can only be signed in incremental mode.',
             SignPdfError.TYPE_PARSE,
         );
     }
 
+    // if (root.indexOf('AcroForm') !== -1) {
```
Comment (vbuch) [inline src/helpers/plainAddPlaceholder/readPdf.js]: No commented out code, please

File: src/helpers/plainAddPlaceholder/readPdf.test.js [hunk 5]
```diff
@@ -3,28 +3,16 @@ import readPdf from './readPdf';
...
-        try {
-            readPdf(pdf);
-            expect('here').not.toBe('here');
-        } catch (e) {
-            expect(e instanceof SignPdfError).toBe(true);
-            expect(e.type).toBe(SignPdfError.TYPE_PARSE);
-            expect(e.message).toMatchSnapshot();
-        }
-    });
     it('Errors when a PDF that already contains forms is passed in', () => {
```
Comment (vbuch) [inline src/helpers/plainAddPlaceholder/readPdf.test.js]: This test is no longer valid as the functionallity should be available. Remove the whole test.

File: src/helpers/plainAddPlaceholder/readRefTable.test.js [hunk 6]
```diff
@@ -3,16 +3,16 @@ import readRefTable from './readRefTable';
 import SignPdfError from '../../SignPdfError';
 
 describe('readRefTable', () => {
-    it('Errors when wrong ref table position is given', () => {
-        const pdf = fs.readFileSync(`${__dirname}/../../../resources/signed.pdf`);
+    it('should be tested', () => {
+        // const pdf = fs.readFileSync(`${__dirname}/../../../resources/signed.pdf`);
```
Comment (vbuch) [inline src/helpers/plainAddPlaceholder/readRefTable.test.js]: It should indeed be tested.

File: src/helpers/removeTrailingNewLine.js [hunk 7]
```diff
@@ -24,10 +24,10 @@ const removeTrailingNewLine = (pdf) => {
 
     const lastLine = output.slice(output.length - 6).toString();
     if (lastLine !== '\n%%EOF') {
-        throw new SignPdfError(
-            'A PDF file must end with an EOF line.',
-            SignPdfError.TYPE_PARSE,
-        );
+        output = Buffer.concat([
```
Comment (vbuch) [inline src/helpers/removeTrailingNewLine.js]: Not quite sure about that. We don't want to alter the document by appending EOF when one is not present. Or do we?

File: src/helpers/removeTrailingNewLine.test.js [hunk 8]
```diff
@@ -11,16 +11,6 @@ describe('removeTrailingNewLine', () => {
             expect(e.type).toBe(SignPdfError.TYPE_INPUT);
         }
     });
-    it('errors if there is no EOF', () => {
```
Comment (vbuch) [inline src/helpers/removeTrailingNewLine.test.js]: I still think we should be checking for that.

File: src/signpdf.test.js [hunk 9]
```diff
@@ -147,6 +148,25 @@ describe('Test signing', () => {
...
     });
+    it('signs a ready pdf two times', async () => {
+        const p12Buffer = fs.readFileSync(`${__dirname}/../resources/certificate.p12`);
+        let pdfBuffer = fs.readFileSync(`${__dirname}/../resources/w3dummy.pdf`);
+        pdfBuffer = plainAddPlaceholder({
+            pdfBuffer, 
+            reason: 'first'
+        });
+        pdfBuffer = signer.sign(pdfBuffer, p12Buffer);
+        fs.writeFileSync(`${__dirname}/../resources/signed-once.pdf`, pdfBuffer);
```
Comment (vbuch) [inline src/signpdf.test.js]: Writing to fs is OK while developing, but we don't want it in CI. So maybe just store a signed.pdf in resources and try to apply a sectond signature.
Ideally signatures would be verified with a devDependency (@MohammedEssehemy has the right thing for this), but that is a separate thing. Verifying however and whatever we can here is OK.

File: src/helpers/pdfkitAddPlaceholder.js [hunk 10]
```diff
@@ -29,28 +31,61 @@ const pdfkitAddPlaceholder = ({
...
+    }
+    const signatureName = 'Signature';
+
     // Generate signature annotation widget
     const widget = pdf.ref({
         Type: 'Annot',
         Subtype: 'Widget',
         FT: 'Sig',
-        Rect: [0, 0, 0, 0],
+        Rect: [188.67, 679.07, 384.78, 748.04],
```
Comment (vbuch) [inline src/helpers/pdfkitAddPlaceholder.js]: This should somehow be configurable or left as it used to be (as an invisible signature)

File: src/helpers/plainAddPlaceholder/getPagesDictionaryRef.js (hunk 1 above)
Comment (therpobinski) [reply src/helpers/plainAddPlaceholder/getPagesDictionaryRef.js]: Excellent, it is not only the regular expression but the way of handling the error. That would solve everything and cover all unit tests. I will make a new branch at the fork of @andres-blanco by uploading the solution and requesting a mix to the develop branch.
Thank you.

File: src/signpdf.test.js (hunk 9 above)
Comment (MohammedEssehemy) [reply src/signpdf.test.js]: We have made some progress in [verify-pdf](https://github.com/ninja-labs-tech/verify-pdf/) hopefully this will help

Comment (andres-blanco) [reply src/signpdf.test.js]: Hey @MohammedEssehemy ! I tried to use verify-pdf but when I ran it over a pdf with two signatures I get this output:
``` 
    console.log src/signpdf.test.js:162
      { verified: false,
        authenticity: false,
        integrity: true,
        expired: false,
        meta:
         { certs: [ [Object] ],
           signatureMeta:
            { reason: 'first',
              contactInfo: 'emailfromp1289@gmail.com',
              location: 'Location from p12' } } }
```
It doesn't seem to recognize the second signature.
If I run mutools I get this output:
```
➜  node-signpdf git:(develop) ✗ mutool sign -v resources/signed-twice.pdf                   
verifying signature 18
  Designated name: CN=signpdf, O=Some Company, OU=, emailAddress=signpdf@example.com, C=BG
  Certificate error: Self-signed certificate.
  The signature is valid but there have been edits since signing.
verifying signature 21
  Designated name: CN=SignPdf, O=node-signpdf, OU=, emailAddress=node-signpdf@example.com, C=BG
  Certificate error: Self-signed certificate.
  The document is unchanged since signing.
```
The file is this:
[signed-twice.pdf](https://github.com/vbuch/node-signpdf/files/3810645/signed-twice.pdf)

**Comment by andres-blanco** [comment]
I've made some modifications, I am working on getting the coverage back to 100%

File: src/helpers/plainAddPlaceholder/readRefTable.test.js (hunk 6 above)
Comment (vizicsaba89) [reply src/helpers/plainAddPlaceholder/readRefTable.test.js]: You're right @andres-blanco

**Comment by vbuch** [comment]
Nice! I'll give it a rry tomorrow and if I don't have remarks, will merge

File: src/signpdf.test.js (hunk 9 above)
Comment (MohammedEssehemy) [reply src/signpdf.test.js]: We are still working on adding support for multi signature @andres-blanco

**Comment by vizicsaba89** [comment]
Lately i didn't have much time, so thank you guys for fixing these bugs / typos. As soon as i have available time, i'll start working on adding configurable images and texts to digital signatures.

File: src/signpdf.test.js (hunk 9 above)
Comment (vbuch) [reply src/signpdf.test.js]: Either way, this is a "good to have" but not a mandatory check. As long as we do some testing.

File: src/helpers/plainAddPlaceholder/readRefTable.test.js (hunk 6 above)
Comment (vbuch) [reply src/helpers/plainAddPlaceholder/readRefTable.test.js]: No. What I meant with "It should indeed be tested" is that the readRefTable function needs some testing. I know it is covered, but coverage != testing. A readRefTable.test.js file is needed that covers at least some edge cases.

**Comment by andres-blanco** [comment]
> I think there is only one thing that needs to be done here before it gets merged. That is: I want to see some tests of the readRefTable function. This would be mostly "in-depth code review" with some written code. Some testing. Nothing too much in detail. Maybe just snapshot testing of the current `resources`? Just making sure there are no regressions in future. Whatever you guys think would suffice.

Great, I'll see what I can do. I think tomorrow I will be able to sit down with this.

**Comment by therpobinski** [comment]
Hi guys, I'm sorry to bother you again, I made a fork from [SIU-Toba/node-signpdf](https://github.com/SIU-Toba/node-signpdf) where I worked a little of what I needed, but when occupying my repository forked the archives of the /dist folder are different from the files in the /src folder, I understand that when using as a dependency the /dist files are used.
How could I do so I can use the last modified code?
I have performed the following steps to use my fork as a dependency, 

1. npm rund build, on my fork after having made the modifications and made a push, to create the /dist folder but it still does not work.
2. I added as a dependency to my project: "node-signpdf": "therpobinski / node-signpdf"
3. npm install, in my project.

When using this dependency I still have the problem that it does not allow signing more than once. I got confused with that.
I hope you can help me, thank you very much.

**Comment by andres-blanco** [comment]
I've added some basic tests to prevent regressions on readRefTable. 
I wont be able to write deeper tests on this function since I am not that familiar with the xref table of pdf files.
Let me know if you think this is enough

**Comment by vbuch** [comment]
I just reused your test @andres-blanco and took a snapshot of the ref tables of all the pdfs available in the resources dir. I think that will save us from regressions. I will merge this in the develop branch of vbuch/node-signpdf but won't release it yet. Will do that later today or tomorrow because I need to check something first.

Thank you all for the contribution! @andres-blanco @therpobinski @vizicsaba89

File: src/signpdf.test.js (hunk 9 above)
Comment (MohammedEssehemy) [reply src/signpdf.test.js]: I've created an issue here
https://github.com/ninja-labs-tech/verify-pdf/issues/8
and will try working on it, once we have try free time.
